server/ai_live_video.go

```diff
@@ -138,5 +139,35 @@ func startControlPublish(control *url.URL, params aiRequestParams) {
 	}
 	params.node.LiveMu.Lock()
 	defer params.node.LiveMu.Unlock()
-	params.node.LivePipelines[stream] = &core.LivePipeline{ControlPub: controlPub}
+
+	ticker := time.NewTicker(10 * time.Second)
+	done := make(chan bool, 1)
+	stop := func() {
+		ticker.Stop()
+		done <- true
+	}
+
+	params.node.LivePipelines[stream] = &core.LivePipeline{
+		ControlPub:  controlPub,
+		StopControl: stop,
+	}
+
+	// send a keepalive periodically to keep both ends of the connection alive
```

Comment:
Slightly confused by this. I thought trickle didn't hold any long running connections? It just individually POST'ed or GET'ed segments from the server. Or is it that it works with a "long get" in case there are no messages to be received? If that's the case, the problem could be solved on a protocol level too, making the long GET stop with a "try again" after those 10s, so we don't need to reimplement this on every channel that doesn't get messages all the time (e.g. I'm creating the events channel now. it naturally has some recurring events, but it could potentially not have too).

Reply:
> Or is it that it works with a "long get" in case there are no messages to be received?

Yep that is exactly what it does.

> If that's the case, the problem could be solved on a protocol level too, making the long GET stop with a "try again" after those 10s

This is actually how the golang subscriber does it already. I agree handling it on the protocol level is better but it's also a bit tricky to do correctly, and we would have to re-implement similar logic both in pytrickle and on the publisher side too (to avoid the server from sweeping idle streams).

That behavior will be good to implement across all the clients in the medium term, but for now this is a decent 80/20 solution.

>  I'm creating the events channel now. it naturally has some recurring events, but it could potentially not have too

There might be some other informative telemetry we can send back periodically for events, eg in/out frame rates ... but if not, then yeah a keepalive would help for now.

Golang subscriber code:

https://github.com/livepeer/go-livepeer/blob/0e1c676707c57d1022f4003b4a8aae25ac1d5116/trickle/trickle_subscriber.go#L86-L120